test/boost/statement_restrictions_test.cc

```diff
@@ -359,3 +359,136 @@ SEASTAR_TEST_CASE(slice_single_column_mixed_order) {
                     singular({I(1), I(1)}), singular({I(1), I(2)})}));
     });
 }
+
+// Currently expression doesn't have operator==().
+// Implementing it is ugly, because there are shared pointers and the term base class.
+// For testing purposes checking stringified expressions is enough.
+static bool expression_eq(const expr::expression& e1, const expr::expression& e2) {
+    return to_string(e1) == to_string(e2);
+}
+
+static void assert_expr_vec_eq(const std::vector<expr::expression>& v1, const std::vector<expr::expression>& v2) {
+    if (std::equal(v1.begin(), v1.end(), v2.begin(), v2.end(), expression_eq)) {
+        return;
+    }
+
+    BOOST_FAIL(fmt::format("Expression vectors not equal! [{}] != [{}]", fmt::join(v1, ", "), fmt::join(v2, ", ")));
+}
+
+// Unit tests for extract_column_restrictions function
+BOOST_AUTO_TEST_CASE(expression_extract_column_restrictions) {
+    using namespace expr;
+
+    auto make_column = [](const char* name, column_kind kind, int id) -> column_definition {
+        column_definition definition(name, int32_type, kind, id);
+
+        ::shared_ptr<column_identifier> identifier = ::make_shared<column_identifier>(name, true);
+        column_specification specification("ks", "cf", std::move(identifier), int32_type);
+        definition.column_specification = std::move(specification);
+
+        return definition;
+    };
+
+    column_definition col_pk1 = make_column("pk1", column_kind::partition_key, 0);
+    column_definition col_pk2 = make_column("pk2", column_kind::partition_key, 1);
+    column_definition col_ck1 = make_column("ck1", column_kind::clustering_key, 0);
+    column_definition col_ck2 = make_column("ck2", column_kind::clustering_key, 1);
+    column_definition col_r1 = make_column("r2", column_kind::regular_column, 0);
+    column_definition col_r2 = make_column("r2", column_kind::regular_column, 1);
+    column_definition col_r3 = make_column("r3", column_kind::regular_column, 2);
+
+    // Empty input test
+    assert_expr_vec_eq(extract_column_restrictions(conjunction{}, col_pk1), {});
```

Comment:
Why can't you use BOOST_TEST instead?

Reply:
Maybe it would be possible, but it would require a custom comparator, and the [documentation](https://www.boost.org/doc/libs/1_76_0/libs/test/doc/html/boost_test/utf_reference/testing_tool_ref/assertion_boost_test_universal_macro.html) for it is nonexistent.

I can try doing it, but I'm not sure if it's worth it.

Reply:
> Maybe it would be possible, but it would require a custom comparator

Would it, really?  I thought it defaulted to operator==, which you provide...

Reply:
I added 
```c++
static bool operator==(const std::vector<expr::expression>& v1, const std::vector<expr::expression>& v2) {
    return std::equal(v1.begin(), v1.end(), v2.begin(), v2.end(), expression_eq);
}

static bool operator==(const expr::expression& v1, const expr::expression& v2) {
    return expression_eq(v1, v2);
}
```
To provide `operator==`

And then did:
```c++
std::vector<expression> a;
BOOST_TEST(a == a);
```

This fails with a long error message
[err.txt](https://github.com/scylladb/scylla/files/6886307/err.txt)

It looks like it tries to use the default variant `operator==`, but fails because it's elements don't have `operator==` :/

Reply:
Yeah, I tried the same thing on my machine (as well as some other things), and I couldn't find a handy way to use BOOST_* macros inside the testcase itself.  Just keep it the way it is now.